spring-cloud-starter-stream-source-sftp/src/test/java/org/springframework/cloud/stream/app/sftp/source/SftpSourceIntegrationTests.java

```diff
@@ -170,10 +177,48 @@ public void streamSourceFilesAsContents() throws InterruptedException {
 
 	}
 
+	@TestPropertySource(properties = { "sftp.listOnly = true",
+			"sftp.factory.host = 127.0.0.1",
+			"sftp.factory.username = user",
+			"sftp.factory.password = pass",
+			"sftp.metadata.redis.keyName = sftpSourceTest" })
+	public static class SftpListOnlyGatewayTests extends SftpSourceIntegrationTests {
+		@Value("${sftp.metadata.redis.keyName}")
+		private String keyName;
+
+		@Before
+		public void before() {
+			redisTemplate.delete(keyName);
+		}
+
+		@Test
+		public void listFiles() throws InterruptedException {
+			for (int i = 1; i <= 2; i++) {
+				@SuppressWarnings("unchecked")
+				Message<String> received = (Message<String>) this.messageCollector.forChannel(sftpSource.output())
+						.poll(10, TimeUnit.SECONDS);
+
+				assertNotNull("No files were received", received);
+
+				assertNotNull("Payload is null", received.getPayload());
+				String filename = received.getPayload();
+				assertEquals("Unexpected payload value", "sftpSource/sftpSource" + i + ".txt", filename);
+			}
+
+			String file1 = "sftpSource/sftpSource1.txt";
+			String file2 = "sftpSource/sftpSource2.txt";
+
+			Thread.sleep(2000); // wait for redis to catch up
```

Comment:
Well, would be better to iterate with short sleep interval and poll Redis for expectations.

This way we won't always wait for 2 seconds for nothing, plus there is no guarantee that really 2 secs are enough.
Our CI servers confirms that the best time is about 10 seconds. But still not always. 

Nevertheless it's not clear to me why Redis wouldn't have the data which is already received from the output channel.

Neither `IdempotentReceiverInterceptor`, nor `RedisMetadataStore` is async. Is it done in the Redis client async way?
What made you to have this sleep here?

Thanks

Reply:
I won't accept the PR until I get an answer or polishing here.

Thanks

Reply:
haven't ignored your question here.. your mention of async and the behavior i was seeing with writes and reads made me interested in finding out more of the root cause. when i have an answer, ill make an update

Reply:
Below are my more in depth findings on this issue. A delete was executed in a jUnit `@Before` method to start with a clean slate. A `Thread.sleep` was introduced prior to fetching entries as there were for the most part consistent failures in which only one record was retrieved for asserting on when there should have been two. 

After further debugging I found that in failure cases the ordering of commands, specifically `DEL` were received by Redis out of order.

For example, a successful run of the test case utilizing a `Thread.sleep` and delete in the @Before method would result in the following proper sequence of commands:

```
1520389387.718441 [0 127.0.0.1:52055] "DEL" "sftpSourceTest"
1520389387.720254 [0 127.0.0.1:52055] "HSETNX" "sftpSourceTest" "sftpSource/sftpSource1.txt" "1520389387648"
1520389387.734632 [0 127.0.0.1:52055] "HSETNX" "sftpSourceTest" "sftpSource/sftpSource2.txt" "1520389387733"
1520389387.735974 [0 127.0.0.1:52055] "HGETALL" "sftpSourceTest"
```

`DEL` is my test case deleting entries before the test method runs, `HSETNX` commands are sent via `@IdempotentReceiver` functionality and `HGETALL` was called in my test case after reading from the `MessageCollector` was finished to assert on the number of values stored in Redis.

When removing the `Thread.sleep` but keeping the delete in the `@Before` method, a failing run of the test case would result in the following sequence of commands:

```
1520389411.523622 [0 127.0.0.1:52082] "HSETNX" "sftpSourceTest" "sftpSource/sftpSource1.txt" "1520389411463"
1520389411.526178 [0 127.0.0.1:52082] "DEL" "sftpSourceTest"
1520389411.539051 [0 127.0.0.1:52082] "HGET" "sftpSourceTest" "sftpSource/sftpSource1.txt"
1520389411.543840 [0 127.0.0.1:52082] "HSETNX" "sftpSourceTest" "sftpSource/sftpSource2.txt" "1520389411542"
1520389411.545270 [0 127.0.0.1:52082] "HGETALL" "sftpSourceTest"
```

As you can see `DEL` is no longer the first command, yet `HSETNX` followed by `DEL` followed by `HGET` on a non-existant entry, followed by `HSETNX` for the second entry. `HGETALL` was then called from the test and the assertion would fail as there was only one entry rather than two.

I have removed the `Thread.sleep` call and moved the delete call into a jUnit `@After` method to cleanup after the test has run. For whatever reason this keeps proper ordering consistently and the tests pass. The sequence of commands appear to always be in proper order and data gets cleaned up:

```
1520397839.919147 [0 127.0.0.1:56725] "HSETNX" "sftpSourceTest" "sftpSource/sftpSource1.txt" "1520397839913"
1520397839.933841 [0 127.0.0.1:56725] "HSETNX" "sftpSourceTest" "sftpSource/sftpSource2.txt" "1520397839933"
1520397839.936068 [0 127.0.0.1:56725] "HGETALL" "sftpSourceTest"
1520397839.937158 [0 127.0.0.1:56725] "DEL" "sftpSourceTest"
```

I have executed a series of 50 consecutive builds and none has failed. Not being a Redis expert, I cannot explain much further than the above. For the time being using `@After` to do cleanup seems reasonable and I don't think there is much value in spending any further debugging time on this.

Reply:
That was great investigation and observation, @chrisjs !

Thank you very much!

Now looking to your finding I can tell what and how happens there.

The ``SftpSourceIntegrationTests` is an application context based test-case and it definitely starts it *before* running any tests. And in our case even before that `@Before`. Since our `SftpSourceApplication` Boot configuration picks up a `SourcePollingChannelAdapter` to read from SFTP and starts it in the end of the the application context. And that one is good to poll files and store metadata in Redis via `IdempotentReceiver`. And right, one polling cycle may happen even before that `@Before`. That's how it's already may be late to perform `DEL` because data has been read and write already. Clear race condition when we deal with multi-threaded situation.

So, I think so far your solution with the `@After` is good enough to clean the testing data after test suit.

:+1:

Reply:
great, thanks for the confirmation. i had suspicions about the context loading and so on when i started noticing random jumps out of the `@Before` method.